docs/algorithms/synthetic_preference_dataset.md

```diff
@@ -0,0 +1,81 @@
+# Synthetic preference dataset
+
+This section focuses explicitly on creating synthetic preference datasets.
+
+# Debug run (use an interactive session)
+
+This code supports HF models, local models and also API-based models (e.g., `gpt-4`). For generating completions, the code now accepts one model at a time, but we're working on adding an ensemble of models. Stay tuned. 
+
+```bash
+# 1. first sample a bunch of completions given prompts
+# Here is an example created dataset: https://huggingface.co/datasets/vwxyzjn/generation_1725567768
+python open_instruct/rejection_sampling/generation.py \
+    --dataset_mixer_list HuggingFaceH4/no_robots 100 \
+    --dataset_splits train \
+    --model_name_or_path allenai/llama-3-tulu-2-8b \
+    --num_completions 3 \
+    --save_filename output/completions.jsonl \
+    --sanity_check \
+    --push_to_hub
+```
+
+### Create preference pairs
+
+```bash
+# 2.1 do LLM as a judge to create synthetic preference dataset
+# Here is an example created dataset: https://huggingface.co/datasets/vwxyzjn/synthetic_preference_dataset_1725567862
+python open_instruct/rejection_sampling/synthetic_preference_dataset.py \
+    --input_filename output/completions.jsonl \
+    --model gpt-4o-2024-08-06 \
+    --save_filename output/synthetic_preferences.jsonl \
+    --num_completions 3 \
+    --push_to_hub \
+```
+
+
+You can visualize the dataset via 
+
+```bash
+python -m costa_utils.hf_viz \
+    --sft vwxyzjn/synthetic_preference_dataset_1725567862 \
+    --split train \
+    --sft_messages_column_name whole_conversation
+
+python -m costa_utils.hf_viz \
+    --preference vwxyzjn/synthetic_preference_dataset_1725567862 \
+    --split train \
+    --preference_chosen_column_name chosen \
+    --preference_rejected_column_name rejected
+```
+
+![synthetic_preference_dataset](synthetic_preference_dataset.png)
+
+# Run through the entire dataset run
+
+To run through the entire dataset you would need a lot more GPUs to finish the generation more quickly. 
+
+
+```bash
+# NOTE: the scripts below only generate 400 prompts, so it's for demonstration purposes only. The scripts are highly scalable, and you could modify its `num_prompts=400` to something else like 300000 for the tulu dataset.
+
+# you need to make sure your default beaker workspace has WANDB_API_KEY and HF_TOKEN secrets in them
```

Comment:
Should this also include BEAKER_TOKEN and OPENAI_API_KEY?

Reply:
Those are included via `mason.py`

https://github.com/allenai/open-instruct/blob/2dbae58f851c9bff2b5ad08fe9a3de588af2e254/mason.py#L123-L128